src/backend/access/nbtree/nbtinsert.c

```diff
@@ -1981,7 +1984,31 @@ _bt_split(Relation rel, Relation heaprel, BTScanInsert itup_key, Buffer buf,
 		XLogBeginInsert();
 		XLogRegisterData((char *) &xlrec, SizeOfBtreeSplit);
 
-		XLogRegisterBuffer(0, buf, REGBUF_STANDARD);
+		/*
+		 * NEON: If we split to earlier pages during a btree vacuum cycle,
+		 * then we have to include the cycle ID in the WAL record. The
+		 * easiest method to do that is to force an image, which happens to
+		 * be relatively cheap, as the data already contained in the record is
+		 * enough to populate the new right page.
+		 *
+		 * We MUST log an FPI when:
+		 *  - The right page's blckno < the left page's blckno
+		 *  - The right page might be 'C' in a page spit chain B > C > A after
+		 *    B split B > A => B > C > A; or B > C > D > A, etc. (as indicated
+		 *    by the presense of a cycle ID).
+		 */
+		if (lopaque->btpo_cycleid == 0 || (rightpagenumber > origpagenumber &&
+										   lopaque->btpo_cycleid != old_cycleid))
+		{
+			/* no cycle ID is required */
+			XLogRegisterBuffer(0, buf, REGBUF_STANDARD);
+		}
+		else
+		{
+			/* cycle ID is required */
+			XLogRegisterBuffer(0, buf, REGBUF_FORCE_IMAGE | REGBUF_STANDARD);
+		}
```

Comment:
Ok, I understand it now. I had to walk through an example to convince myself it's correct:

1. Initially, there are three pages in the leaf level, forming a chain like this:

```
    9 <-> 10 <-> 20
```

2. VACUUM begins and processes page 9

3. Page 10 is split into pages 10 and 5. Because rightpage < leftpage,  this sets the cycle ID on both pages, which is depicted with *. 

```
    9 <-> 10* <-> 5* <-> 20
```

4. Page 5 is split further into pages 5 and 6. The cycle ID is set on both pages, because it was already set on the left page. 
```
    9 <-> 10* <-> 5* <-> 6* <-> 20
```

5. Vacuum continues on page 10. It follows the right links to also process 5 and 6 because the cycle IDs are set.

It's a bit awkward that we have essentially the same backtracking logic here that we have in btvacuumpage() in upstream, but formulated slightly differently. I'm not sure I have any better ideas though. We could skip these optimizations and always WAL-log the cycle ID if it's non-zero, but that's not very appetizing either. Maybe just try to improve the comments a bit. A reference to the corresponding logic in btvacuumpage() would be good for example.

Looking at the logic in btvacuumpage(), we can also skip this on internal pages; the cycle id only matter on leaf pages. And I'm still a bit uneasy about the BTP_SPLIT_END flag.

To make this easier to test, we could move the logic earlier, and clear the cycle ID in the in-memory page too, if the conditions are not met. That way, you'd get the same behaviour even if the pages fit in cache.

Reply:
Ok. Let's just comments to explain why that's OK, then.